Add `strict` option

schinkowitch opened the pull request:
Changes the contract of the throttle from never executing more than limit times within an interval to never waiting less than the interval between limit executions.

Fixes #26

sindresorhus commented on test.js:
```diff
@@ -21,6 +20,38 @@ test('main', async t => {
+
+	await throttled(0);
+
+	await new Promise(resolve => {
```
Just use the `delay` dependency.

Richienb replied:
Use `yoctodelay` instead.

schinkowitch replied:
Used delay

sindresorhus commented on test.js:
```diff
@@ -21,6 +20,38 @@ test('main', async t => {
+
+	const results = await Promise.all(promises);
+
+	results.forEach((executed, index) => {
```
Use a `for-of` loop (just my preference).

schinkowitch replied:
Ok

sindresorhus commented on index.js:
```diff
@@ -17,32 +17,24 @@ const pThrottle = ({limit, interval}) => {
 				};
 
 				const now = Date.now();
+				const executeAt = ticks.length < limit ? now : ticks[limit - 1] + interval;
```
```suggestion
				const runDate = ticks.length < limit ? now : ticks[limit - 1] + interval;
```

might be a clearer name.

schinkowitch replied:
I changed the algorithm a bit so this is no longer relevant.

sindresorhus commented:
Thanks for working on this fix? Would you be able to add some more tests? I'm a little bit paranoid about changing the implementation this much, so would be good to ensure corner cases are covered.

schinkowitch commented:
I updated this pull request, so that the current algorithm is used by default. I added a strict option to use the strict throttling algorithm. I updated this algorithm compared to the previous commit, so that it only stores "limit" execution times instead of storing the execution times of all the queued calls, when using the strict algorithm.

schinkowitch commented:
Merged changes and fixed merge error.

schinkowitch commented:
Hi again.

I made a new package schinkowitch/strict-throttle that takes the ideas of the changes in this pull request and implements them with some improvements. So, feel free to close this without merging, but the code is yours if you wish to use it.

sindresorhus commented:
Thanks :)

Sorry about the slow response. I just had too many PRs to review.
